File: src/main/kotlin/sunsetsatellite/vm/sunlite/SLFunction.kt
package sunsetsatellite.vm.sunlite

import sunsetsatellite.lang.sunlite.FunctionModifier
import sunsetsatellite.lang.sunlite.Param
import sunsetsatellite.lang.sunlite.Type

class SLFunction(val name: String,
                 val returnType: Type,
                 val params: List<Param>,
                 val typeParams: List<Param>,
                 val chunk: Chunk,
                 val arity: Int = 0,
                 val upvalueCount: Int,
                 val localsCount: Int,
                 val modifier: FunctionModifier = FunctionModifier.NORMAL
) {

	override fun toString(): String {
		return "<function '${name}(${params.map { it.type }.joinToString()}): ${returnType}'>"
	}

	fun copy(): SLFunction {
		return SLFunction(name, returnType, params, typeParams, chunk, arity, upvalueCount, localsCount, modifier)
	}
}

File: src/main/csharp/SunliteSharp/Runtime/SLFunction.cs
using SunliteSharp.Core;
using SunliteSharp.Core.Modifier;
using Type = SunliteSharp.Core.Type;

namespace SunliteSharp.Runtime;

public record SLFunction(
    string Name,
    Type ReturnType,
    List<Param> Params,
    List<Param> TypeParams,
    Chunk Chunk,
    int Arity,
    int UpvalueCount,
    int LocalsCount,
    FunctionModifier Modifier
)
{
    public override string ToString()
    {
        return $"<function '{Name}({string.Join(", ",Params.Select( p => p.Type))}): {ReturnType}'>";
    }

    public SLFunction Copy()
    {
        return new SLFunction(Name, ReturnType, Params, TypeParams, Chunk, Arity, UpvalueCount, LocalsCount, Modifier);
    }
};